### PR TITLE
Shine Adjustments and Jump Cancel Boma Extension

### LINKED ISSUE #670: Shine on block Frame Advantage Readjustment
With the removal of character's being unable to grab during blockstun, a few moves have to readjust in terms of their frame advantage:

New Proposed Down Special Frame Advantage:

Fox - +1
Wolf/Dpit - +2
Falco/Zelda - +3
Lucas/Ness - +4

Old Down Special Frame Advantage:

Fox/Wolf/Lucas/Ness - +1
Dpit - +2
Zelda - +3
Falco - +4

First off, the main determining factor of the shine frame advantage is that shine to jc grab will always lose to roll, so no shine like move will ever be above +4. 

- Next, is a combination of accessibility and jump squat. Fox/Wolf/Falco shines are ordered in this fashion due to the differences in jump squat and reward off of grab. Generally fox usually has the best reward, then wolf, and then falco. 

- Dpit is left in the +2 (could honestly be +1/0, we're most likely gonna remove his shine at some point as it isn't necessary in his tool kit) as his shine comes out slower than the rest at frame 5.

- Zelda follows a similar logic to Dpit and she is placed in the +3 category as her shine comes out frame 8. She does have some benefits in comparison to the others as her shine does have a significantly larger hitbox along with intangibility on frames 5-8.

- Lucas magnet is frame 5 like dpit so one would think it makes sense for him to be at +2 like Dpit. However, unlike the other characters on the list, he has a tether grab that is active on frame 14 so the threat of grabbing from his magnet is not nearly as frame tight as the other characters thus affording him leeway on his frame advantage. We also have to take into consideration that he is a DJC character, so a lot of his pressure comes from doing magnet -> djc aerial stuff which is very strong but he wouldn't be able to straight up grab you from it. So he has more pressure when doing an aerial magnet to DJC aerial. Which is a pretty central point to how he was designed in PM and I'm assuming  that is in the same vein of the design direction HDR is taking with the character. 

- With Ness, I'm very reluctant to place him at +4. While you can avoid the grab with a frame perfect roll (in reality with tap buffer, the window is 5 frames), it would be reasonable to give the defender a two frame leeway in avoiding the grab in these scenarios. The reason Ness was given the +4 was due to his magnet being slower than the rest at 7 frames. Also among the characters in the list, he would most likely have the hardest time actually getting magnet on shield against his opponents due to the combination of his speed and aerial frame data. If his magnet were to be improved to frame 5, then the case can be made to make it +3. But as it stands, it doesn't make much sense to put it in the same frame advantage range as the other characters whose shines are generally much more applicable for shield pressure.

### PR DESCRIPTION
**Pull request by MadVillainLordQ**
Resolves #670

Notes:
- Added a jump cancel boma extension which makes it much easier to reference a jump cancel function for future usage.
- Lucas shine frame advantage can't be changed until me (or another dev) can figure out how to make his jump cancel occur one frame sooner than current.
- Some moves such as Bowser Jr. and Yoshi Side B will most likely need adjustments in the future. 
- Wolf and Falco Shine activity has been adjusted to 1 frame

[Shine PR Changelog.txt](https://github.com/HDR-Development/HewDraw-Remix/files/9142520/Shine.PR.Changelog.txt)

### CODE REVIEW

**Comment by yoonikoo** [comment]
The shine adjustments to both Wolf and Falco both fit PM and/or Melee. Good readjustments to fit their former iterations.

File: fighters/fox/src/opff.rs [hunk 1]
```diff
@@ -18,22 +18,31 @@ unsafe fn laser_fastfall_landcancel(boma: &mut BattleObjectModuleAccessor, statu
 }
 
 // Fox Shine Jump Cancels
-unsafe fn shine_jump_cancel(boma: &mut BattleObjectModuleAccessor, status_kind: i32, situation_kind: i32, cat1: i32) {
-    if boma.is_status(*FIGHTER_STATUS_KIND_SPECIAL_LW) && WorkModule::get_int(boma, *FIGHTER_INSTANCE_WORK_ID_INT_FRAME_IN_AIR) <= 1 {
-        GroundModule::correct(boma, app::GroundCorrectKind(*GROUND_CORRECT_KIND_GROUND));
```
Comment (jobrien97) [inline fighters/fox/src/opff.rs:23]: You got rid of multishine functionality here

File: fighters/wolf/src/opff.rs [hunk 2]
```diff
@@ -18,25 +18,33 @@ unsafe fn airdodge_cancel(boma: &mut BattleObjectModuleAccessor, status_kind: i3
 }
 
 // Wolf Shine Jump Cancels
-unsafe fn shine_jump_cancel(boma: &mut BattleObjectModuleAccessor, status_kind: i32, situation_kind: i32, cat1: i32) {
-    if boma.is_status(*FIGHTER_STATUS_KIND_SPECIAL_LW) && WorkModule::get_int(boma, *FIGHTER_INSTANCE_WORK_ID_INT_FRAME_IN_AIR) <= 1 {
-        GroundModule::correct(boma, app::GroundCorrectKind(*GROUND_CORRECT_KIND_GROUND));
```
Comment (jobrien97) [inline fighters/wolf/src/opff.rs:23]: You got rid of multishine functionality here too

File: fighters/fox/src/opff.rs [hunk 3]
```diff
@@ -18,22 +18,31 @@ unsafe fn laser_fastfall_landcancel(boma: &mut BattleObjectModuleAccessor, statu
...
+        can_jump_cancel = false;
+    }
+    if can_jump_cancel {
+        WorkModule::enable_transition_term(fighter.module_accessor, *FIGHTER_STATUS_TRANSITION_TERM_ID_CONT_JUMP_SQUAT);
+        WorkModule::enable_transition_term(fighter.module_accessor, *FIGHTER_STATUS_TRANSITION_TERM_ID_CONT_JUMP_SQUAT_BUTTON);
+        WorkModule::enable_transition_term(fighter.module_accessor, *FIGHTER_STATUS_TRANSITION_TERM_ID_CONT_JUMP_AERIAL);
+        WorkModule::enable_transition_term(fighter.module_accessor, *FIGHTER_STATUS_TRANSITION_TERM_ID_CONT_JUMP_AERIAL_BUTTON);
+        fighter.sub_transition_group_check_ground_jump_mini_attack();
+        fighter.sub_transition_group_check_ground_jump();
+        fighter.sub_transition_group_check_air_jump_aerial();
```
Comment (jobrien97) [inline fighters/fox/src/opff.rs:45]: Why aren't you just using `is_input_jump` boma ext and directly transitioning to jumpsquat/double jump status based on situation kind?

**Review by jobrien97 — CHANGES_REQUESTED** [review]
I don't think a can_jump_cancel bool is necessary here. It'd be cleaner to call the jump functions directly for those two cases

File: fighters/fox/src/acmd/specials.rs [hunk 4]
```diff
@@ -105,6 +105,8 @@ unsafe fn game_speciallwstart(fighter: &mut L2CAgentBase) {
         ATTACK(fighter, 0, 0, Hash40::new("top"), 3.0, 360, 32, 0, 62, 8.0, 0.0, 6.5, 0.0, None, None, None, 0.9, 1.0, *ATTACK_SETOFF_KIND_OFF, *ATTACK_LR_CHECK_POS, false, -1, 0.0, 0, false, false, false, false, true, *COLLISION_SITUATION_MASK_G, *COLLISION_CATEGORY_MASK_ALL, *COLLISION_PART_MASK_ALL, false, Hash40::new("collision_attr_elec"), *ATTACK_SOUND_LEVEL_S, *COLLISION_SOUND_ATTR_FIRE, *ATTACK_REGION_ENERGY);
         ATTACK(fighter, 1, 0, Hash40::new("top"), 3.0, 24, 45, 0, 66, 8.0, 0.0, 6.5, 0.0, None, None, None, 0.9, 1.0, *ATTACK_SETOFF_KIND_OFF, *ATTACK_LR_CHECK_POS, false, -1, 0.0, 0, false, false, false, false, true, *COLLISION_SITUATION_MASK_A, *COLLISION_CATEGORY_MASK_ALL, *COLLISION_PART_MASK_ALL, false, Hash40::new("collision_attr_elec"), *ATTACK_SOUND_LEVEL_S, *COLLISION_SOUND_ATTR_FIRE, *ATTACK_REGION_ENERGY);
         AttackModule::set_add_reaction_frame(boma, 0, 2.0, false);
+        ReflectorModule::set_status(boma, *FIGHTER_FOX_REFLECTOR_KIND_REFLECTOR, app::ShieldStatus(*SHIELD_STATUS_NORMAL), *FIGHTER_REFLECTOR_GROUP_EXTEND);
```
Comment (jobrien97) [inline fighters/fox/src/acmd/specials.rs:108]: The reflect frame might be a param in vl.prc, not sure

Comment (MadVillainLordQ) [reply fighters/fox/src/acmd/specials.rs:108]: Tried to use the only relevant vl param of reflector_attack_frame, and that did not work. Not sure what else to do about this one.

File: fighters/wolf/src/opff.rs (hunk 2 above)
Comment (MadVillainLordQ) [reply fighters/wolf/src/opff.rs:23]: Rectified this issue

File: fighters/fox/src/opff.rs (hunk 1 above)
Comment (MadVillainLordQ) [reply fighters/fox/src/opff.rs:23]: Rectified this issue

File: fighters/falco/src/opff.rs [hunk 5]
```diff
@@ -21,40 +21,48 @@ unsafe fn laser_ff_land_cancel(boma: &mut BattleObjectModuleAccessor, status_kin
...
+                fighter.jump_cancel();
             }
-        }
+            else if fighter.is_status_one_of (&[
+                *FIGHTER_FOX_STATUS_KIND_SPECIAL_LW_HIT,
+                *FIGHTER_FOX_STATUS_KIND_SPECIAL_LW_LOOP,
+                *FIGHTER_FOX_STATUS_KIND_SPECIAL_LW_END
+                ]) && !fighter.is_in_hitlag() {
+                fighter.jump_cancel();
+            }
```
Comment (jobrien97) [inline fighters/falco/src/opff.rs]: Make this an OR block please

File: fighters/falco/src/opff.rs [hunk 6]
```diff
@@ -21,40 +21,48 @@ unsafe fn laser_ff_land_cancel(boma: &mut BattleObjectModuleAccessor, status_kin
...
+            else {
+                false;
+            }
```
Comment (jobrien97) [inline fighters/falco/src/opff.rs]: Not needed anymore

File: dynamic/src/ext.rs [hunk 7]
```diff
@@ -783,6 +786,20 @@ impl BomaExt for BattleObjectModuleAccessor {
         return StatusModule::status_kind(self);
     }
 
+    unsafe fn jump_cancel(&mut self) {
```
Comment (jobrien97) [inline dynamic/src/ext.rs]: change to check_jump_cancel pls

**Review by jobrien97 — CHANGES_REQUESTED** [review]

**Review by jobrien97 — APPROVED** [review]
lgtm, nice work

**Pull request merged**